roles/repositories/tasks/main.yml

```diff
@@ -0,0 +1,74 @@
+---
+- name: 'Enable Red Hat Repositories'
+  theforeman.foreman.repository_set:
+    username: "{{ username }}"
+    password: "{{ password }}"
+    server_url: "{{ server_url }}"
+    validate_certs: "{{ validate_certs | default(omit) }}"
+    organization: "{{ organization }}"
+    product: "{{ item.0.name }}"
+    name: "{{ item.1.name }}"
+    state: enabled
+    repositories:
+      - releasever: "{{ item.1.releasever | default(omit) }}"
+        basearch: "{{ item.1.basearch | default(omit) }}"
+  with_subelements:
+    - "{{ products }}"
+    - repository_sets
+
+- name: 'Create Products'
+  theforeman.foreman.product:
+    username: "{{ username }}"
+    password: "{{ password }}"
+    server_url: "{{ server_url }}"
+    validate_certs: "{{ validate_certs | default(omit) }}"
+    organization: "{{ organization }}"
+    name: "{{ item.name }}"
+    state: present
+  with_items:
+    - "{{ products | selectattr('repositories', 'defined') | list }}"
+
+- name: 'Create Repositories'
+  theforeman.foreman.repository:
+    username: "{{ username }}"
+    password: "{{ password }}"
+    server_url: "{{ server_url }}"
+    validate_certs: "{{ validate_certs | default(omit) }}"
+    organization: "{{ organization }}"
+    name: "{{ item.1.name }}"
+    url: "{{ item.1.url }}"
+    product: "{{ item.0.name }}"
+    content_type: "{{ item.1.content_type }}"
+    deb_components: "{{ item.1.deb_components | default(omit) }}"
+    deb_architectures: "{{ item.1.deb_architectures | default(omit) }}"
+    deb_releases: "{{ item.1.deb_releases | default(omit) }}"
+    download_policy: "{{ item.1.download_policy | default('on demand') }}"
+    mirror_on_sync: "{{ item.1.mirror_on_sync | default(true) }}"
+    state: present
+  with_subelements:
+    - "{{ products | selectattr('repositories', 'defined') | list }}"
+    - repositories
+
+- name: 'Create Sync Plans'
```

Comment:
Is there a benefit to putting this inside the repositories role instead of split out into it's own role? I would think the latter provides more composibility for users.

Reply:
The benefit is that defining it in the same role allows us to specify that a repository should be synced only when changed.

I'm still debating whether that's worth doing, as well as researching the data structures involved when registering the results of the with_subelements loops above.

Reply:
Ah, I thought this comment was relevent to Syncs, not Sync Plans. My apologies. Please disregard my prior comment.

Reply:
The benefit here is that the user can define data for their Products alongside data for their Sync Plans.

Regarding composibility, I've taken care to ensure that each piece should be optional and tasks should be skipped when the required data structure is not defined.

@ehelms do you think there would be value in defining this role to be re-usable so that it can be passed multiple times with different variables for any users who desire that approach?

Reply:
I'm just having a hard time seeing the value provided by putting this all in a single role versus providing users more flexibility by having roles for configuring repositories, sync plans and taking the action to sync a repository. I think `sync` is a great example of this. Not all users will want to sync their repositories when they create or enable them. Users will want to sync repositories out of band of enable and creation. Therefore, we will end up with a stand-alone sync role anyway. So I perceive it as better to just write the stand-alone sync role (it could always be included by the repositories role).

Reply:
I think at some point I expressed the desire of having a "meta" data structure that defines "all things content" and all content related roles feed from the same structure, to avoid duplication of information in similar but not identical structures. This might be influenced by that idea?

Reply:
Yes. One goal I have is that the user should never have to define the same data in multiple places.

That's still possible with when composing role, but what it looks like is a single composite role, which receives the data once, and passes it to the child roles.

IMO at that point it's easier and more maintainable to just have a single role

Reply:
> That's still possible with when composing role, but what it looks like is a single composite role, which receives the data once, and passes it to the child roles.

I do not really understand this nuance. But I do not feel my point was addressed either. A dedicated sync role gives more power to users and allows use in this case.

A user will definitely have a common data structure spanning these roles. The repositories structure might be used by a user when defining their content views for example.

I think we are in that weird space between providing "workflows" and providing "components" for use by a user. My concern is if a user does not want to initiate a sync everytime they change something (add a repository, or product, change a property of a repository) they have to incur a sync. And maybe they prefer to let sync plans drive that, maybe their CI/CD but the control is taken away from them. Whereas a stand alone sync role allows the user to construct playbooks that define their workflow.

Reply:
After thinking more on this, I agree, sync plans should go into an own role.